src/backends/pytorch/libtorch_backend.cc

```diff
@@ -0,0 +1,745 @@
+    buffer_copy_offset += expected_byte_size;
+  }
+
+  RETURN_IF_ERROR(SetInputTensor(
+      name, shape, dtype, static_cast<char*>(buffer), total_byte_size));
+  return Status::Success;
+}
+
+Status
+LibTorchBackend::Context::SetInputTensor(
+    const std::string& name, const std::vector<int64_t>& shape,
+    const DataType dtype, char* content, size_t byte_size)
+{
+  const auto pr = ConvertDataTypeToTorchType(dtype);
+  if (!pr.first) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Failed to convert DataType '" +
+                                         DataType_Name(dtype) +
+                                         "' to Torch datatype");
+  }
+
+  // torch::TensorOptions options =
+  //     torch::TensorOptions().device(device_).dtype(pr.second);
+  torch::Tensor input_tensor = torch::from_blob(content, shape, pr.second);
+  input_tensor = input_tensor.to(device_);
+
+  if (input_tensor.nbytes() != byte_size) {
+    return Status(
+        RequestStatusCode::INTERNAL,
+        "unexpected size " + std::to_string(byte_size) +
+            " for inference input '" + name + "', expecting " +
+            std::to_string(input_tensor.nbytes()));
+  }
+  inputs_.push_back(input_tensor);
+
+  return Status::Success;
+}
+
+Status
+LibTorchBackend::Context::ReadFixedSizedOutputTensor(
+    const std::string& name, const int& op_index, const DataType dtype, const size_t dtype_byte_size,
+    const size_t total_batch_size, std::vector<Scheduler::Payload>* payloads)
+{
+  std::vector<int64_t> content_shape;
+  char* content = nullptr;
+  size_t byte_size = 0;
+  RETURN_IF_ERROR(
+      GetOutputTensor(name, op_index, dtype, &content, &byte_size, &content_shape));
+
+  const size_t total_byte_size =
+      GetElementCount(content_shape) * dtype_byte_size;
+  const size_t batch1_byte_size = total_byte_size / total_batch_size;
+
+  if (byte_size != total_byte_size) {
+    return Status(
+        RequestStatusCode::INTERNAL,
+        "unexpected size for output '" + name + "', byte-size " +
+            std::to_string(byte_size) + " does not equal " +
+            std::to_string(total_batch_size) + " * " +
+            std::to_string(batch1_byte_size));
+  }
+
+  size_t content_offset = 0;
+
+  for (auto& payload : *payloads) {
+    const InferRequestHeader& request_header =
+        payload.request_provider_->RequestHeader();
+    const size_t expected_byte_size =
+        request_header.batch_size() * batch1_byte_size;
+
+    // If 'payload' requested this output then copy it from
+    // 'content'. If it did not request this output then just
+    // skip it in the 'content'.
+    if ((payload.response_provider_ != nullptr) &&
+        payload.response_provider_->RequiresOutput(name)) {
+      void* buffer;
+      Status status = payload.response_provider_->AllocateOutputBuffer(
+          name, &buffer, expected_byte_size, content_shape);
+      if (status.IsOk()) {
+        memcpy(buffer, content + content_offset, expected_byte_size);
```

Comment:
Extra memcpy. You copy once from output tensor to `content`, and then from `content` to `buffer`, should be able to copy directly from output tensor to `buffer`

Reply:
There still seems to be an additional copy in GetOutputTensor...

Reply:
Removed extra memcpy